overlord/devicestate/systems.go

```diff
@@ -0,0 +1,153 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2020 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package devicestate
+
+import (
+	"fmt"
+
+	"github.com/snapcore/snapd/dirs"
+	"github.com/snapcore/snapd/overlord/snapstate"
+	"github.com/snapcore/snapd/overlord/state"
+	"github.com/snapcore/snapd/seed"
+)
+
+func checkSystemRequestConflict(st *state.State, systemLabel string) error {
+	st.Lock()
+	defer st.Unlock()
+
+	var seeded bool
+	if err := st.Get("seeded", &seeded); err != nil && err != state.ErrNoState {
+		return err
+	}
+	if seeded {
+		// the system is fully seeded already
+		return nil
+	}
+
+	// inspect the current system which is stored in modeenv, were are
+	// holding the state lock so there is no race against mark-seeded
+	// clearing recovery system; recovery system is not cleared when seeding
+	// fails
+	modeEnv, err := maybeReadModeenv()
+	if err != nil {
+		return err
+	}
+	if modeEnv == nil {
+		// non UC20 systems do not support actions, no conflict can
+		// happen
+		return nil
+	}
+
+	// not yet fully seeded, hold off requests for the system that is being
+	// seeded, but allow requests for other systems
+	if modeEnv.RecoverySystem == systemLabel {
+		return &snapstate.ChangeConflictError{Message: "cannot request system action, system is seeding"}
```

Comment:
we probably want to set ChangeKind: "seed" as other similar cases

Reply:
Added